crates/nargo/src/cli/mod.rs

```diff
@@ -206,6 +207,37 @@ fn path_to_stdlib() -> PathBuf {
     dirs::config_dir().unwrap().join("noir-lang").join("std/src")
 }
 
+// Removes duplicates from the list of public input witnesses
+fn dedup_public_input_indices(indices: PublicInputs) -> PublicInputs {
+    let duplicates_removed: HashSet<_> = indices.0.into_iter().collect();
+    PublicInputs(duplicates_removed.into_iter().collect())
+}
+
+// Removes duplicates from the list of public input witnesses and the
+// associated list of duplicate values.
+pub(crate) fn dedup_public_input_indices_values(
+    indices: PublicInputs,
+    values: Vec<FieldElement>,
+) -> (PublicInputs, Vec<FieldElement>) {
+    // Assume that the public input index lists and the values contain duplicates
+    assert_eq!(indices.0.len(), values.len());
+
+    let mut public_inputs_without_duplicates = Vec::new();
+    let mut already_seen_public_indices = HashSet::new();
+
+    for (index, value) in indices.0.iter().zip(values) {
+        if !already_seen_public_indices.contains(&index) {
+            already_seen_public_indices.insert(index);
+            public_inputs_without_duplicates.push(value)
+        }
+    }
+
+    (
+        PublicInputs(already_seen_public_indices.into_iter().cloned().collect()),
```

Comment:
This would be an optimisation, whereas we haven't benchmarked to see if we really need to that

Reply:
I don't think we need a benchmark, swap_remove is O(1), so it will be O(d), where d is the number of duplicates among the inputs.
In your case it will be O(n) where n is the number of inputs since you need to copy the set into a vector.

Reply:
The current method is more readable and there is nothing to say that its too slow, so there is no need to optimise it